ambition_utils/rrule/models.py

```diff
@@ -324,27 +328,44 @@ def generate_dates(self, num_dates=20):
         dates = []
 
         rule = self.get_rrule()
+        
+        def append_date(date):
+            """
+            Appends date, in UTC, to list if it satisfies start_date logic.
+            """
+            date_with_tz = fleming.attach_tz_if_none(date, self.time_zone)
+            date_in_utc = fleming.convert_to_tz(date_with_tz, pytz.utc, True)
+
+            if not start_date or date_in_utc > start_date:
+                dates.append(date_in_utc)
 
         try:
+            # Capture, and append, the rule's first date for .after() considering in the loop.
             d = rule[0]
-            # Convert to time zone
-            date_with_tz = fleming.attach_tz_if_none(d, self.time_zone)
-            date_in_utc = fleming.convert_to_tz(date_with_tz, pytz.utc, True)
-            dates.append(date_in_utc)
+            append_date(d)
```

Comment:
i am not a big fan of this nested function that is modifying a list outside of the function and checking other variables defined outside of the function. is it possible to turn this `append_date` function to a method that does the time zone conversion and returns a date? then you could have a validation method to determine if it should be appended.

seems like it would be easier to write tests around the different cases in the logic + make it more straightforward when reading the code.

something like
```
d = rule[0]
date_in_utc = self.convert_to_utc(d, self.time_zone)
if self.date_is_in_range(date_in_utc, start_date):
    dates.append(date_in_utc)
```

Reply:
Good suggestions. `RRule.convert_to_utc` already exists so I opted to use it.

That leaves the range logic. I don't like a method outside of `get_dates()` since this _very_ specific to `get_dates()`. I'm OK with nested method but opted for maximum readability by having the basic condition in both places.

I tried to improve readability by improving comments, improving variable names, separating some lines.

Reply:
Tests are still passing when ran locally.